### ISSUE TITLE
(zap.c:579:zap_deref_leaf()) VERIFY3(zap->zap_u.zap_fat.zap_phys->zap_magic == 0x2F52AB2ABULL)

### ISSUE THREAD

**Issue by chrisrd** [zfsonlinux/zfs#1445]
linux-3.8.10, zfsonlinux/spl@352bd19, zfsonlinux/zfs@57f5a20

The system was largely unresponsive (very high load,  following a large number of "remove xattr, unlink file" sequences on a` xattr=dir` file system, ref: #457), eventually leading to a power cycle. Following this, I have one file system producing this when trying to mount:

```
May  7 08:00:43 b5 kernel: [  817.900053] SPLError: 15166:0:(zap.c:579:zap_deref_leaf()) VERIFY3(zap->zap_u.zap_fat.zap_phys->zap_magic == 0x2F52AB2ABULL) failed (0 == 12703150763)
May  7 08:00:43 b5 kernel: [  817.900148] SPLError: 15166:0:(zap.c:579:zap_deref_leaf()) SPL PANIC
May  7 08:00:43 b5 kernel: [  817.900206] SPL: Showing stack for process 15166
May  7 08:00:43 b5 kernel: [  817.900259] Pid: 15166, comm: mount.zfs Tainted: G           O 3.8.8-otn-00007-g5914d52 #1
May  7 08:00:43 b5 kernel: [  817.900323] Call Trace:
May  7 08:00:43 b5 kernel: [  817.900387]  [<ffffffffa0e45557>] spl_debug_dumpstack+0x27/0x40 [spl]
May  7 08:00:43 b5 kernel: [  817.900449]  [<ffffffffa0e466cf>] spl_debug_bug+0x7f/0xe0 [spl]
May  7 08:00:43 b5 kernel: [  817.900548]  [<ffffffffa07b8251>] zap_deref_leaf+0xa1/0x1e0 [zfs]
May  7 08:00:43 b5 kernel: [  817.900632]  [<ffffffffa07bf6cd>] ? zap_lockdir+0x12d/0xd80 [zfs]
May  7 08:00:43 b5 kernel: [  817.900713]  [<ffffffffa0784086>] ? sa_lookup+0x66/0x120 [zfs]
May  7 08:00:43 b5 kernel: [  817.900796]  [<ffffffffa07baac1>] fzap_cursor_retrieve+0xe1/0x3f0 [zfs]
May  7 08:00:43 b5 kernel: [  817.900881]  [<ffffffffa07c079b>] zap_cursor_retrieve+0x6b/0x450 [zfs]
May  7 08:00:43 b5 kernel: [  817.900968]  [<ffffffffa07f1204>] ? zfs_znode_alloc+0x4f4/0x7f0 [zfs]
May  7 08:00:43 b5 kernel: [  817.901052]  [<ffffffffa07cc90e>] zfs_purgedir+0xae/0x270 [zfs]
May  7 08:00:43 b5 kernel: [  817.901136]  [<ffffffffa07f461d>] ? zfs_zget+0x39d/0x3d0 [zfs]
May  7 08:00:43 b5 kernel: [  817.901218]  [<ffffffffa07ccbce>] zfs_unlinked_drain+0xfe/0x150 [zfs]
May  7 08:00:43 b5 kernel: [  817.901282]  [<ffffffffa0e50009>] ? taskq_create+0x309/0x4c0 [spl]
May  7 08:00:43 b5 kernel: [  817.901362]  [<ffffffffa076e4c4>] ? dsl_prop_register+0x454/0x560 [zfs]
May  7 08:00:43 b5 kernel: [  817.901424]  [<ffffffff81063d70>] ? finish_wait+0x80/0x80
May  7 08:00:43 b5 kernel: [  817.901529]  [<ffffffffa07eebb0>] ? zfs_rmdir+0x9f0/0x9f0 [zfs]
May  7 08:00:43 b5 kernel: [  817.901636]  [<ffffffffa07fb70f>] ? zil_open+0x6f/0x160 [zfs]
May  7 08:00:43 b5 kernel: [  817.901742]  [<ffffffffa07e2ae0>] zfs_sb_setup+0x120/0x1c0 [zfs]
May  7 08:00:43 b5 kernel: [  817.901848]  [<ffffffffa07e46f0>] zfs_domount+0x240/0x2e0 [zfs]
May  7 08:00:43 b5 kernel: [  817.901954]  [<ffffffffa080deb0>] ? zpl_mount+0x30/0x30 [zfs]
May  7 08:00:43 b5 kernel: [  817.902059]  [<ffffffffa080dec3>] zpl_fill_super+0x13/0x80 [zfs]
May  7 08:00:43 b5 kernel: [  817.902165]  [<ffffffffa080deb0>] ? zpl_mount+0x30/0x30 [zfs]
May  7 08:00:43 b5 kernel: [  817.902248]  [<ffffffff8113e686>] mount_nodev+0x56/0xb0
May  7 08:00:43 b5 kernel: [  817.902352]  [<ffffffffa080dea5>] zpl_mount+0x25/0x30 [zfs]
May  7 08:00:43 b5 kernel: [  817.902494]  [<ffffffff8113f5c0>] mount_fs+0x20/0xd0
May  7 08:00:43 b5 kernel: [  817.902576]  [<ffffffff81159174>] vfs_kern_mount+0x74/0x110
May  7 08:00:43 b5 kernel: [  817.902657]  [<ffffffff8115b697>] do_mount+0x357/0xa60
May  7 08:00:43 b5 kernel: [  817.902737]  [<ffffffff810e620e>] ? __get_free_pages+0xe/0x50
May  7 08:00:43 b5 kernel: [  817.902817]  [<ffffffff8115b1aa>] ? copy_mount_options+0x3a/0x180
May  7 08:00:43 b5 kernel: [  817.902899]  [<ffffffff8115be2e>] sys_mount+0x8e/0xe0
May  7 08:00:43 b5 kernel: [  817.902980]  [<ffffffff813a2e59>] system_call_fastpath+0x16/0x1b
```

A similar problem was seen in #1242, see 2nd trace.

**Comment by ryao** [comment]
@behlendorf This just occurred on my workstation yesterday evening during system updates. I am archiving a disk image of the root pool before attempting to repair it. I can test any patches that are aimed at fixing this on a copy of the image.

**Comment by behlendorf** [comment]
This issue was resolved by commit 29572ccdeff7ddb1211b0f26dea69e0a2f262faf.
